src/gui/vector/qgsvectorlayerproperties.cpp

```diff
@@ -1792,6 +1861,73 @@ void QgsVectorLayerProperties::pbnUpdateExtents_clicked()
   mMetadataFilled = false;
 }
 
+void QgsVectorLayerProperties::pbnDcUnlock_clicked()
+{
+  cbxUseEstimatedMetadata->setEnabled( true );
+  cbxSelectAtId->setEnabled( true );
+  cbxCheckPrimaryKeyUnicity->setEnabled( true );
+  pbnDcCancel->setEnabled( true );
+  pbnDcApply->setEnabled( true );
+  pbnDcUnlock->setEnabled( false );
+}
+
+void QgsVectorLayerProperties::pbnDcCancel_clicked()
+{
+  const QgsDataSourceUri uri = mLayer->dataProvider()->uri();
+  QgsVectorDataProvider::Capabilities capabilities = mLayer->dataProvider()->capabilities();
+  // Datasource primary key disable select at id
+  if ( capabilities & QgsVectorDataProvider::CanDisableSelectAtId )
+  {
+    cbxSelectAtId->setChecked( !uri.selectAtIdDisabled() );
+  }
+  // Datasource primary key disable check unicity
+  if ( capabilities & QgsVectorDataProvider::CanDisbaleCheckPrimaryKeyUnicity )
+  {
+    const QString checkUnicityKey { QStringLiteral( "checkPrimaryKeyUnicity" ) };
+    cbxUseEstimatedMetadata->setChecked( uri.hasParam( checkUnicityKey ) &&
+                                         uri.param( checkUnicityKey ).compare( QLatin1String( "1" ) ) == 0 );
+  }
+  // Datasource use estimated metadata capability
+  if ( capabilities & QgsVectorDataProvider::CanUseEstimatedMetadata )
+  {
+    cbxUseEstimatedMetadata->setChecked( uri.useEstimatedMetadata() );
+  }
+
+  cbxUseEstimatedMetadata->setEnabled( false );
+  cbxSelectAtId->setEnabled( false );
+  cbxCheckPrimaryKeyUnicity->setEnabled( false );
+  pbnDcUnlock->setEnabled( true );
+  pbnDcCancel->setEnabled( false );
+  pbnDcApply->setEnabled( false );
+}
+
+void QgsVectorLayerProperties::pbnDcApply_clicked()
```

Comment:
Why are we using an explicit Apply option for this setting? Couldn't we just use the dialog's existing "Apply" button instead?

Reply:
I thought that the application of changes in the layer provider's data source should be done separately but without a new dialog.

Reply:
I think users would find this a bit odd. @nirvn what do you think?

Reply:
I'm -1 to have a separate apply settings button here. I can't see what would justify the confusing "nested" applies UX.

Reply:
@nyalldawson @nirvn I have change the buttons like. `Cancel` becomes `Reset` and `Apply` becomes `Update` datasource. It is the same as `Update extent`, the `Update datasource` changes a core layer data.